include/deal.II/lac/generic_linear_algebra.h

```diff
@@ -154,10 +229,68 @@ namespace LinearAlgebraTrilinos
 {
   using namespace dealii;
 
+  /**
+   * Typedef for the base class for preconditioners.
+   *
+   * This typedef is useful when wanting to create a generic solver interface
+   * wherein one can switch between preconditioners. For example:
+   * @code
+   * std_cxx11::unique_ptr<LA::PreconditionerBase> preconditioner;
+   * if (parameters.preconditioner_type == "Jacobi") {
+   * LA::MPI::PreconditionJacobi* ptr_prec
+   *   = new LA::MPI::PreconditionJacobi ();
+   * ...
+   * preconditioner.reset(ptr_prec);
```

Comment:
To initialise the preconditioner and assign AdditionalData to it, the compiler needs to know its datatype. So

```c++
std_cxx11::unique_ptr<TrilinosWrappers::PreconditionBase> preconditioner; // I want this to be able to make a run-time choice of preconditioner
preconditioner.reset(new LA::MPI::PreconditionJacobi());
LA::MPI::PreconditionJacobi::AdditionalData
  additional_data (parameters.preconditioner_relaxation);

preconditioner->initialize(tangent_matrix.block(u_block,u_block), additional_data);
``` 
results in
```bash
error: no member named 'initialize' in 'dealii::TrilinosWrappers::PreconditionBase'
            preconditioner->initialize(tangent_matrix.block(u_block,u_block),
```

I'll put a full example in the docs.